### DISCUSSION TITLE
[API Request] Endpoint to fetch song list (getSongList2)

### DISCUSSION THREAD

**Discussion by jeffvli** [opensubsonic/open-subsonic-api#18] [API requests]
### Type of change

API extension

### Proposal description

Tangentially related to #6

Add a dedicated song list endpoint similar to `getAlbumList2`. There is currently no way for users to directly browse through their songs except for using the `search3` endpoint (which does not support filtering).

### Backward compatibility

- [X] No backward compatibility impact.

### API details

To follow existing API naming standards, the endpoint should be named: `getSongList2`.

Starting off, parameters that are included in `getAlbumList2` should also be included here. An additional parameter called `byAlbum` should be included to sort the resulting output by their respective albums instead of the song directly.

```
GET /getSongList2

Parameters:
- type (required)
  - random
  - newest
  - highest
  - frequent
  - recent
  - starred
  - genre
  - fromYear
  - toYear
  - musicFolderId
  - byAlbum
  - byYear
  - byGenre
  - alphabeticalByName
  - alphabeticalByArtist
- size -- number (max 500)
- offset -- pagination offset
- fromYear (required if type is byYear) -- number
- toYear (required if type is byYear) -- number
- genre (required if type is byGenre) -- name of genre
```

`byAlbum` example response:
```
Without byAlbum, the album will be ignored and the list parameters will be queried directly on the songs
GET /getSongList2?type=alphabeticalByName

{
  "subsonic-response": {
    "status": "ok",
    "version": "1.16.1",
    "type": "AwesomeServerName",
    "serverVersion": "0.1.3 (tag)",
    "songList2": {
      "song": [
        {
          "name": "AAA",
          "album": "DDD",
          ...rest of song response
        },
        {
          "name": "BBB",
          "album": "LLL",
          ...rest of song response
        },
        {
          "name": "CCC",
          "album": "DDD",
          ...rest of song response
        }
      ]
    }
  }
}


With byAlbum, the sort is performed on the albums instead which will the return the song list grouped by the album
GET /getSongList2?type=alphabeticalByName&byAlbum=true

{
  "subsonic-response": {
    "status": "ok",
    "version": "1.16.1",
    "type": "AwesomeServerName",
    "serverVersion": "0.1.3 (tag)",
    "songList2": {
      "song": [
        {
          "name": "ZZZ (does not matter)",
          "album": "AAA"
          ...rest of song response
        },
        {
          "name": "BBB (does not matter)",
          "album": "AAA"
          ...rest of song response
        },
        {
          "name": "AAA (does not matter)",
          "album": "CCC"
          ...rest of song response
        }
      ]
    }
  }
}
```

Also add additional parameter `includeTotalRecordCount` from #16 to return the total record count.

Example response:
```
GET /getSongList2?type=newest&size=2&offset=0&includeTotalRecordCount=true

{
  "subsonic-response": {
    "status": "ok",
    "version": "1.16.1",
    "type": "AwesomeServerName",
    "serverVersion": "0.1.3 (tag)",
    "songList2": {
      "totalRecordCount": 2,
      "song": [
        {
          ...song 1 response
        },
        {
          ...song 2 response
        }
      ]
    }
  }
}
```

**Comment by Tolriq** [comment]
What does ` - byAlbum` do ?

`  - alphabeticalByName` Not a fan to have order defined like that this should be the default without type.
Alphabetical can mean quite a few things, there's unicode sort, ascii sort and then I hope in the future the API will support sort title tags. So that's another alpha sort. IMO the sort should be another parameter for those case and of course there's asc/desc issue.
`  - alphabeticalByArtist` What does it mean? Should have an id for the artist? Again I do not like the sort in the type it should be split.
And if it's an all songs but sorted by artist name then I'm against this as it will block support for multiple artists supports. (And it should go in the no type filter and some new sort value like displayArtist that will be more clear about what artist it's sorted by).

**Reply by jeffvli** [reply]
Now that I think about it some more, `byAlbum` probably doesn't make sense. This would rather be an `type=album` sort, where songs are sorted in ascending order by their album title, while other types such as `type=genre` would automatically group albums together where applicable.

>   - alphabeticalByName Not a fan to have order defined like that this should be the default without type.
Alphabetical can mean quite a few things, there's unicode sort, ascii sort and then I hope in the future the API will support sort title tags. So that's another alpha sort. IMO the sort should be another parameter for those case and of course there's asc/desc issue.
>  - alphabeticalByArtist What does it mean? Should have an id for the artist? Again I do not like the sort in the type it should be split.
And if it's an all songs but sorted by artist name then I'm against this as it will block support for multiple artists supports. (And it should go in the no type filter and some new sort value like displayArtist that will be more clear about what artist it's sorted by).

The idea was to structure the endpoint to be mostly identical to the existing `getAlbumList2` endpoint for ease of implementation by server owners. The `type` parameter is required in `getAlbumList2`, so you would have to use `type=alphabeticalByName` for an A-Z sort.  I'm not sure how much we want to deviate from the existing endpoints but I'm open to updating both this as well as creating a new `getAlbumList3` or similar endpoint if we want to have it fit more modern standards with improved sort and filtering.

**Reply by Tolriq** [reply]
IMO new endpoints should be modern endpoints ready to embrace the future. The goal of the extensions versus a 2.0 is also to easy the transition.
With a nice new endpoint in the extension this will also help to have other nice new end points for the rest and even without realizing it in a couple of years we have a new proper API that would never have existed if we did it one shot from the start with most servers not having the BW to handle that migration.
TL;DR; The first new endpoints should be modern and thought enough to serve as a base of what could have been a 2.0. (So clearly defined query and result without polymorphism)

**Reply by jeffvli** [reply]
Sounds good to me!

I was under the impression that server owners chose extensions since any new endpoints would mostly conform to the existing Subsonic API standards.

In that case do you think it would be better to just compile a list of API extension ideas and then vote on them before having to do a writeup for the spec? 

I think it might make sense for clients to just describe the endpoint and the functionality they're looking for, and have server owners have the final say on the new API endpoint architecture.

It might make sense to:
1. Open a chat (discord / matrix / etc.) for chat-based discussion regarding some topics (API architecture, questions, etc.)
4. Create a working spreadsheet (Google Sheets) that can be collaborated on in real-time

I don't dislike forum-based discussion but I feel like sometimes it can take a while for ideas and discussions to propagate due to how it's structured. What are your thoughts?

**Comment by Tolriq** [comment]
> I was under the impression that server owners chose extensions since any new endpoints would mostly conform to the existing Subsonic API standards.

From a few discussion with them it's also mostly to keep backward compatibility and not having to directly support a huge new API, but being able to add (or skip) new endpoints on a time / interest basis.

> I think it might make sense for clients to just describe the endpoint and the functionality they're looking for, 

Yes that's what I said in another thread (But it's still important that the ask does provide a attempt at a design to ease the thinking/discussion) , the important first is to define the base brick (Expose that a server support OS and how to get the list of extension please vote in https://github.com/opensubsonic/open-subsonic-api/discussions/5 if not done.

About the tool to compile, there won't be any magic tool fit all. Chat does not really work as everything is mixed up and people just write quickly without taking time to think first.

Google sheet I'm not sure about a proper format to allow proper collaboration and proper mixing, sheets does not really fit the need here.

Once more requests are made I was thinking more about a manual work, to build proper draft extensions.

In the end there won't be that much draft.

I think like 6 / 7 extensions / clarification ones (most of the request are just about adding new fields / data that can be combined in a single draft).
Then there will be a few additions, those will take more time to analyse to try to build a working base anyway.
